File: day02.scala
//> using toolkit 0.6.0

object Day02:
	val path = os.pwd / "input" / "day02" / "example1.txt"
	val lines = os.read.lines(path).toVector
	val reports =
		lines
		.map(_.split(" ").map(_.toInt).toVector)
		.toVector

	def isSafe(report: Vector[Int]): Boolean =
		val increasing = report.indices.drop(1).forall(i => report(i - 1) < report(i))
		val decreasing = report.indices.drop(1).forall(i => report(i - 1) > report(i))
		val diff = report.sliding(2).forall(xs => 
			val d = (xs(0) - xs(1)).abs
			1 <= d && d <= 3
		)
		(increasing || decreasing) && diff

	def partOne(): Unit =
		val result = reports.map(isSafe).count(_ == true)
		println(result)

	def isSafeDampened(report: Vector[Int]): Boolean =
		isSafe(report) || report.indices.exists(i => isSafe(report.patch(i, Nil, 1)))

	def partTwo(): Unit =
		val result = reports.map(isSafeDampened).count(_ == true)
		println(result)

	@main def run(): Unit =
		partOne()
		partTwo()

File: haskell/day02.hs
import System.Environment ( getArgs )

main :: IO ()
main = do
        args <- getArgs
        content <- readFile (head args)
        print (partOne content)
        print (partTwo content)

isAll :: (Int -> Int -> Bool) -> [Int] -> Bool
isAll _ [] = True
isAll _ [x] = True
isAll f (x : xs) = f x (head xs) && isAll f xs

isIncreasing :: [Int] -> Bool
isIncreasing = isAll (<)

isDecreasing :: [Int] -> Bool
isDecreasing = isAll (>)

isCorrectDifference :: [Int] -> Bool
isCorrectDifference = isAll (\x y -> abs (x - y) `elem` [1..3])

isSafe :: [Int] -> Bool
isSafe report =
        (isIncreasing report  ||
         isDecreasing report) &&
        isCorrectDifference report

count :: [Bool] -> Int
count xs = length (filter id xs)

partOne :: String -> Int
partOne input =
        count (map (isSafe . (map read . words)) (lines input))

removeAt :: Int -> [Int] -> [Int]
removeAt i xs =
        case splitAt i xs of
                (xs, ys) -> xs ++ drop 1 ys

isSafeDampenedRecursive :: Int -> [Int] -> Bool
isSafeDampenedRecursive (-1) _ = False
isSafeDampenedRecursive i xs =
        isSafe (removeAt i xs) ||
        isSafeDampenedRecursive (i - 1) xs

isSafeDampened :: [Int] -> Bool
isSafeDampened report =
        isSafe report ||
        isSafeDampenedRecursive (length report) report

partTwo :: String -> Int
partTwo input =
        count (map (isSafeDampened . (map read . words)) (lines input))
